---
title: "BenchmarkST: Cross-Platform, Multi-Species Spatial Transcriptomics Gene Imputation Benchmarking"
# subtitle: "YOUR SUBTITLE (OPTIONAL)"
summary:
authors: [qianru]
#   - USERNAME1
#   - USERNAME2
tags: ["osre24", reproducibility, machine learning, bioinformatics, spatial transcriptomics]
categories: []
date: 2024-06-09
lastmod: 2024-06-09
featured: false
draft: false

# Featured image
# To use, add an image named `featured.jpg/png` to your page's folder.
# Focal points: Smart, Center, TopLeft, Top, TopRight, Left, Right, BottomLeft, Bottom, BottomRight.
image:
  caption: ""
  focal_point: ""
  preview_only: false
---

Hello! My name is Qianru, and I will be working on a project to improve spatial transcriptomics during Google Summer of Code 2024. My project, [Benchmarking Gene Imputation Methods for Spatial Transcriptomics](/project/osre24/uci/benchmarkst/), is mentored by {{% mentione ziheng %}} and [Cormac Flanagan](https://users.soe.ucsc.edu/~cormac/). The goal is to create a standard platform to evaluate methods for filling in missing gene data, which is a big challenge in spatial transcriptomics. [My proposal can be viewed here!](https://drive.google.com/file/d/1ydqGuuzpNgPpVUBvTiFvF1q7qV9gA_wm/view?usp=sharing)

Spatial transcriptomics lets us see where genes are active in tissues, giving us insight into how cells interact in their natural environment. However, current methods often miss some gene data, making it hard to get a complete picture. Gene imputation can help fill in these gaps.

My project will:

Create a benchmark dataset to standardize gene imputation tasks across different platforms, species, and organs.

Compare various gene imputation methods to see how well they work in different scenarios.

Develop a user-friendly Python package with tools for gene imputation to help researchers improve their data.

I'm excited to contribute to this project and help advance the field of spatial transcriptomics by making data analysis more accurate and comprehensive.
